File: recover-binary-search-tree.cpp
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes

#include <iostream>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* first = nullptr;
    TreeNode* last = nullptr;
    TreeNode* prev = nullptr;
    bool flag = false;

    void recoverTree(TreeNode* root) {
        if (root == nullptr) return;

        inorder(root);

        int temp = last->val;
        last->val = first->val;
        first->val = temp;
    }

    void inorder(TreeNode* root) {
        if (root == nullptr) return;

        inorder(root->left);

        if (prev != nullptr && root->val < prev->val) {
            if (!flag) {
                flag = true;
                first = prev;
                last = root;
            } else {
                last = root;
            }
        }

        prev = root;

        inorder(root->right);
    }
};

File: Recover_Binary_Search_Tree.py
# Time Complexity : O(n), where n is the number of nodes in the binary search tree
# Space Complexity : O(h), where h is the height of the binary search tree.

# Definition for a binary tree node.
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def __init__(self):
        self.first = None
        self.second = None
        self.prev = None
    
    def recoverTree(self, root: TreeNode) -> None:
        self.inorderTraversal(root)
        self.first.val, self.second.val = self.second.val, self.first.val
    
    def inorderTraversal(self, node):
        if not node:
            return
        
        self.inorderTraversal(node.left)
        
        if self.prev and self.prev.val > node.val:
            if not self.first:
                self.first = self.prev
            self.second = node
        
        self.prev = node
        
        self.inorderTraversal(node.right)
